### PR TITLE
`mutex.cpp`, `cond.cpp`: Use static dispatch

### PR DESCRIPTION
**Pull request by cpplearner**
After #2317 removed `stl_critical_section_vista` and `stl_condition_variable_vista`, only the Win7 versions inherit from the corresponding interfaces. Thus the functions in `mutex.cpp` and `cond.cpp` could directly use the Win7 classes, without going through the interfaces. This improves performance, and makes it easier to implement `constexpr mutex`.

The first commit (24922f3) changes the getters in `_Mtx_internal_imp_t` and `_Cnd_internal_imp_t` to return the Win7 classes. Since these classes are `final`, apparently the compiler knows to use static dispatch for them.

Subsequent commits change `create_stl_critical_section` and `create_stl_condition_variable` to create non-polymorphic classes, with the "vptr" set to nullptr.

I can imagine that these cause trouble if a `mutex` object is created by new code and used by old code, and old code uses the old function definitions (because it links to an old static lib?). IIUC this scenario is unsupported, and it will be a noticeable breakage due to null dereference.

### CODE REVIEW

**Comment by strega-nil-ms** [comment]
I don't personally like the `stl_critical_section`/`stl_condition_variable` implementation, since I believe it to be unnecessarily dangerous (depending on `reinterpret_cast`-ability between a standard layout type and a type with base classes) without any actual benefit (imo). However, the remaining changes look good.

**Comment by StephanTLavavej** [comment]
Could we try a different approach here? As @strega-nil-ms noted, the reinterpretation is highly squirrelly. Instead, I suggest removing the `meow_interface` base classes, directly using the `meow_win7` classes, and adding the unused `void*` as a vptr placeholder at the beginning.

`CONDITION_VARIABLE_INIT` is cool, I was unaware of that machinery!

**Comment by StephanTLavavej** [comment]
I've pushed a conflict-free merge with `main` to pick up the new toolset (instead of rerunning sporadic failures against the old VM pool).

File: stl/inc/mutex [hunk 1]
```diff
@@ -95,9 +95,8 @@ public:
 
     using native_handle_type = void*;
```
Comment (frederick-vs-ja) [inline stl/inc/mutex:96]: Should we still declare the `native_handle_type` even if `native_handle` is not usable? I guess this is technically conforming, but looks weird.

Comment (StephanTLavavej) [reply stl/inc/mutex:96]: Yeah, it's an open question.

**Comment by AlexGuteniev** [comment]
> makes it easier to implement `constexpr mutex`.

I don't see it.
We need to support old DLL version, which will want vtable.

(I previously attempted a PR with fake vtable instead, that was rejected)

**Comment by cpplearner** [comment]
> > makes it easier to implement `constexpr mutex`.
> 
> I don't see it. We need to support old DLL version, which will want vtable.
> 
> (I previously attempted a PR with fake vtable instead, that was rejected)

Yes, new code will not work with old DLL/LIB, but IIUC that's [unsupported](https://learn.microsoft.com/en-us/cpp/porting/binary-compat-2015-2017#restrictions) anyway.

**Comment by AlexGuteniev** [comment]
> Yes, new code will not work with old DLL/LIB, but IIUC that's [unsupported](https://learn.microsoft.com/en-us/cpp/porting/binary-compat-2015-2017#restrictions) anyway.

I believe it is unsupported for the LIB, but for the DLL it is supported due to "printer driver problem"

**Comment by strega-nil-ms** [comment]
> > makes it easier to implement `constexpr mutex`.
> 
> I don't see it. We need to support old DLL version, which will want vtable.
> 
> (I previously attempted a PR with fake vtable instead, that was rejected)

@AlexGuteniev makes a very good point. If we're working towards `constexpr` `mutex`, we would need to allow the user code to create a `stl_critical_section_win7` with the _correct_ vtable, rather than _no_ vtable. It should theoretically be constexpr to do this if we declare `stl_critical_section_win7` in a public header (although suitably uglified).

**Comment by AlexGuteniev** [comment]
> It should theoretically be constexpr to do this if we declare `stl_critical_section_win7` in a public header (although suitably uglified).

Let's work towards this then!

And then we won't even need manual devirtualization as in this PR: due to `final` keyword the compiler would devirtualize calls.

https://godbolt.org/z/6PM87qvYP

**Comment by strega-nil-ms** [comment]
@AlexGuteniev

ah. That is unfortunate.

**Comment by cpplearner** [comment]
I'd rather not repeat #2309 in this PR. If we decide to preserve the vptr, we can keep only 24922f3 (which alone is a performance improvement) and leave out all subsequent commits.

**Comment by StephanTLavavej** [comment]
@strega-nil-ms
> If we're working towards `constexpr` `mutex`, we would need to allow the user code to create a `stl_critical_section_win7` with the _correct_ vtable, rather than _no_ vtable. It should theoretically be constexpr to do this if we declare `stl_critical_section_win7` in a public header (although suitably uglified).

I don't believe that that's necessary. In fact, I believe that's going in the wrong direction.

Now that I've thought about this code more clearly, I believe that the `stl_critical_section_MEOW` machinery is *almost* entirely confined to the DLL/LIB. That is, the public headers determine how much space/alignment to provide for this machinery, and that's an ABI-immutable fact of life (fortunately, using less space is fine). The public headers also recently started caring where `_Count` appears in the layout, which #3763 fixed (by changing the DLL/LIB layout for UWP to match the public headers' expectations). Aside from that, the public headers almost entirely manipulate this machinery through a flat C API - the `_Mtx_lock()`, `_Cnd_wait()`, etc. families. This is an exceedingly good fact for us. It means that we can change the actual implementation arbitrarily, as long as we don't exceed the space/alignment requirements or move `_Count`.

There is a single exception - we unwisely provided `native_handle()`. However, as I mentioned above, it was not practically usable by users (as we never declared the types that were actually being returned as `void*`). In theory someone could have been calling `native_handle()`, but it wouldn't have achieved anything useful.

Therefore I believe that there are no practical ABI implications to nulling out the vtable. (This might mean that we were unnecessarily paranoid about the earlier "fake vtable" PR that was rejected, but IIRC that PR had extreme complexity that we couldn't effectively reason about, and it was before we boiled away enough of the XP/Vista support machinery to see what was really happening.)

I am also not concerned about the dreaded "print driver" scenario that we may or may not actually need to support (it is unclear whether it is actually guaranteed to work according to our documented restrictions on MSDN). Because this isn't adding new separately compiled APIs (which is the sort of thing that would cause immediate failure on load), and instead would only be detectable upon use of `native_handle()`, I think it falls into the same "effectively unobservable" category.

**Comment by StephanTLavavej** [comment]
Actually, I believe I have convinced myself that we don't even need to retain the `void* unused` null placeholder for the vptr. The headers don't actually care how much of the aligned storage space is used, and (aside from `native_handle()` which I assert no user could feasibly have been calling), nothing cares about the internal layout.

I was led to this by thinking about whether `stl_condition_variable_win7` needed the vptr, given that condition variables have no mirrored `_Count` to worry about. Then I realized that `_Mtx_internal_imp_mirror` / `_Mtx_internal_imp_t` separate the aligned storage (where the `stl_critical_section_win7` lives) from the `_Count` / `count`.

To repeat/clarify, because ABI issues are unfamiliar to reason about, there is only one copy of the STL's DLL or static LIB that we ever need to worry about. (It is possible for multiple user DLLs to be loaded into a single process with separate STL DLLs or statically linked LIBs, but they cannot communicate CRT/STL objects between themselves.) So, the STL's DLL/LIB is always internally consistent and we don't need to worry about mix-and-match the way that we do for headers built by users. The flat C API (with the `_Count` and `native_handle()` "punch through" exceptions as noted) ensures that the user-built headers are insulated from what's happening in the STL's DLL/LIB. As long as `native_handle()` is unused, and `_Count` at the mirrored layer *outside* the aligned storage appears at the expected location, we are fine.

I will push a change. If anyone believes that my analysis is incorrect, I welcome corrections, but please check with me before beginning an "edit war" of commits. :smiley_cat:

**Review by StephanTLavavej — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
One more thought: for a `constexpr mutex` followup PR, we could have an escape hatch, so that "print driver" code could give up `constexpr mutex` and continue calling the separately compiled initialization functions. That would be my preferred path forward, since it would conform by default for the vast majority of code that doesn't care (even among code that takes advantage of bincompat guarantees).

**Comment by AlexGuteniev** [comment]
> I am also not concerned about the dreaded "print driver" scenario that we may or may not actually need to support (it is unclear whether it is actually guaranteed to work according to our documented restrictions on MSDN). Because this isn't adding new separately compiled APIs (which is the sort of thing that would cause immediate failure on load), and instead would only be detectable upon use of `native_handle()`, I think it falls into the same "effectively unobservable" category.

If we don't call the constructor from DLL, and don't initialize vtable by other means, I'm afraid "print driver" scenario would be observable by crash on trying to lock the mutex.

**Comment by AlexGuteniev** [comment]
> One more thought: for a `constexpr mutex` followup PR, we could have an escape hatch, so that "print driver" code could give up `constexpr mutex` and continue calling the separately compiled initialization functions.

My understanding that escape hatch is not good here, unlike, say, atomic alignment escape hatch.
Because the determination whether you affected by "print driver" problem or not is made not when creating app, but by user environment.
Even if technically not claimed supported, I think this scenario should be maximally supported, just because failure to support it would lead to random failures. And in this particular case, when export surface is intact, the failures are more subtle.

**Comment by StephanTLavavej** [comment]
I've pushed a revert, to restore the null vptrs (in advance of the weekly maintainer meeting). @strega-nil-ms noted that a guaranteed crash was safer and better than treating data as a vptr.

We will still need to talk about the print driver scenario for a followup `constexpr mutex` PR - I see your point, @AlexGuteniev, but maximally supporting that scenario is in conflict with many users who have expected and requested `constexpr mutex` conformance. If the scenario does exist in the wild, and people are using `mutex` in it, I don't think the failures will be "random" (like race conditions); they would happen immediately upon attempting to use the `mutex`, as @strega-nil-ms has envisioned by restoring the null vptr.

**Review by StephanTLavavej — APPROVED** [review]

**Review by CaseyCarter — APPROVED** [review]
From Discord discussion, STL plans to address my comments, so I'll go ahead and approve.

**Review by StephanTLavavej — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
I'm mirroring this to the MSVC-internal repo - please notify me if any further changes are pushed.

**Review by CaseyCarter — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
Thanks for this major and long-awaited simplification! :tada: :heart_eyes_cat: :sparkles:

**Pull request merged**